### PR TITLE
WPB-4888: Implement request tracing across federation

### PR DESCRIPTION
**Pull request by battermann**
https://wearezeta.atlassian.net/browse/WPB-4888

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: libs/wire-api-federation/src/Wire/API/Federation/BackendNotifications.hs [hunk 1]
```diff
@@ -31,7 +32,8 @@ data BackendNotification = BackendNotification
     -- this body, which could be very large and completely useless to the
     -- pusher. This also makes development less clunky as we don't have to
     -- create a sum type here for all types of notifications that could exist.
-    body :: RawJson
+    body :: RawJson,
+    requestId :: Maybe RequestId
```
Comment (mdimjasevic) [inline libs/wire-api-federation/src/Wire/API/Federation/BackendNotifications.hs:36]: Why is this an optional field?

File: services/federator/test/unit/Test/Federator/InternalServer.hs [hunk 2]
```diff
@@ -148,7 +148,7 @@ federatedRequestFailureAllowList =
         . runInputConst settings
         . runInputConst (FederationDomainConfigs AllowDynamic [FederationDomainConfig (Domain "hello.world") FullSearch FederationRestrictionAllowAll] 10)
         . interpretMetricsEmpty
-        $ callOutward targetDomain Brig (RPC "get-user-by-handle") request
+        $ callOutward Nothing targetDomain Brig (RPC "get-user-by-handle") request
```
Comment (mdimjasevic) [inline services/federator/test/unit/Test/Federator/InternalServer.hs:151]: Kind of the same as above.

Comment (battermann) [reply services/federator/test/unit/Test/Federator/InternalServer.hs:151]: Also not in scope, or I am missing something.

File: services/federator/test/unit/Test/Federator/Remote.hs [hunk 3]
```diff
@@ -89,7 +89,7 @@ mkTestCall sslCtx hostname port = do
     . discoverLocalhost hostname port
     . assertNoError @DiscoveryFailure
     . interpretRemote
-    $ discoverAndCall (Domain "localhost") Brig "test" [] mempty
+    $ discoverAndCall Nothing (Domain "localhost") Brig "test" [] mempty
```
[the remark this answers is not in the document]
Comment (battermann) [reply services/federator/test/unit/Test/Federator/Remote.hs:92]: It's simply of no use in the unit tests. It's for debugging/tracing requests in prod/staging envs.

File: libs/wire-api-federation/src/Wire/API/Federation/BackendNotifications.hs (hunk 1 above)
Comment (mdimjasevic) [reply libs/wire-api-federation/src/Wire/API/Federation/BackendNotifications.hs:36]: Where is this actually set? I get that in theory when parsing a header it might be missing, but my understanding is we should make it mandatory, i.e., part of every request. If that is the case, then it'd be nice if this can also be reflected in types.

Comment (battermann) [reply libs/wire-api-federation/src/Wire/API/Federation/BackendNotifications.hs:36]: It's set by nginx and it's written into the `Env` as part of the servant middleware.

Comment (mdimjasevic) [reply libs/wire-api-federation/src/Wire/API/Federation/BackendNotifications.hs:36]: OK, I thought so. I'd prefer if we abort in the middleware if there's no request ID, but maybe that's too much.

I'll leave this up to you.

File: services/federator/test/unit/Test/Federator/Remote.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply services/federator/test/unit/Test/Federator/Remote.hs:92]: How about adding a test that asserts there is a request ID?

**Review by mdimjasevic — APPROVED** [review]
It looks fine to me, but I'd recommend adding a test that asserts there is a request ID.

File: libs/wire-api-federation/src/Wire/API/Federation/BackendNotifications.hs (hunk 1 above)
Comment (battermann) [reply libs/wire-api-federation/src/Wire/API/Federation/BackendNotifications.hs:36]: I hear you. The only purpose of the request ID is debugging. If we made it a required header, we would have to change all of our integration tests. IMO this is not justified.

File: libs/wai-utilities/src/Network/Wai/Utilities/Server.hs [hunk 4]
```diff
@@ -361,6 +361,7 @@ onError g m r k e = liftIO $ do
   when (code >= 500) $
     either Prm.incCounter (counterIncr (path "net.errors")) `mapM_` m
   flushRequestBody r
+  Log.flush g
```
Comment (akshaymankar) [inline libs/wai-utilities/src/Network/Wai/Utilities/Server.hs:364]: Why is this necessary?

File: libs/wire-api-federation/src/Wire/API/Federation/BackendNotifications.hs [hunk 5]
```diff
@@ -51,6 +54,7 @@ instance FromJSON BackendNotification where
       <*> o .: "targetComponent"
       <*> o .: "path"
       <*> (RawJson . TL.encodeUtf8 <$> o .: "body")
+      <*> o .: "requestId"
```
Comment (akshaymankar) [inline libs/wire-api-federation/src/Wire/API/Federation/BackendNotifications.hs:57]: This is dangerous. The parser needs to be backwards compatible so any old things which are queued will still get processed after a restart.

File: libs/wire-api-federation/src/Wire/API/Federation/HasNotificationEndpoint.hs [hunk 6]
```diff
@@ -46,10 +47,11 @@ fedNotifToBackendNotif ::
   KnownSymbol (NotificationPath tag) =>
   KnownComponent (NotificationComponent k) =>
   ToJSON (Payload tag) =>
+  RequestId ->
   Domain ->
   Payload tag ->
   BackendNotification
-fedNotifToBackendNotif ownDomain payload =
+fedNotifToBackendNotif mReqId ownDomain payload =
```
Comment (akshaymankar) [inline libs/wire-api-federation/src/Wire/API/Federation/HasNotificationEndpoint.hs:54]:
```suggestion
fedNotifToBackendNotif reqId ownDomain payload =
```

File: libs/wire-api-federation/src/Wire/API/Federation/HasNotificationEndpoint.hs [hunk 7]
```diff
@@ -60,5 +62,6 @@ fedNotifToBackendNotif ownDomain payload =
         { ownDomain = ownDomain,
           targetComponent = componentVal @(NotificationComponent k),
           path = path,
-          body = body
+          body = body,
+          requestId = mReqId
```
Comment (akshaymankar) [inline libs/wire-api-federation/src/Wire/API/Federation/HasNotificationEndpoint.hs:66]:
```suggestion
          requestId = reqId
```

File: services/brig/src/Brig/Run.hs [hunk 8]
```diff
@@ -156,10 +160,15 @@ type ServantCombinedAPI =
...
-  let reqid = maybe def RequestId $ lookupRequestId req
-  mkapp reqid req cont
+lookupRequestIdMiddleware :: Logger -> (RequestId -> Wai.Application) -> Wai.Application
+lookupRequestIdMiddleware logger mkapp req cont = do
+  case lookupRequestId req of
+    Just rid -> do
+      mkapp (RequestId rid) req cont
+    Nothing -> do
+      localRid <- RequestId . cs . UUID.toText <$> UUID.nextRandom
+      Log.info logger $ "request-id" .= localRid ~~ "request" .= (show req) ~~ msg (val "generated a new request id for local request")
```
Comment (akshaymankar) [inline services/brig/src/Brig/Run.hs:170]: I think it would be nicer if log basics of a request like method and path instead of logging the whole request. It could contain sensitive info in headers. This way the logs also become searchable by those things.

File: services/cannon/src/Cannon/Types.hs [hunk 9]
```diff
@@ -120,7 +119,7 @@ runCannon' :: Env -> Cannon a -> IO a
 runCannon' e c = runReaderT (unCannon c) e
 
 lookupReqId :: Request -> RequestId
-lookupReqId = maybe def RequestId . lookup requestIdName . requestHeaders
+lookupReqId = RequestId . fromMaybe "N/A" . lookup requestIdName . requestHeaders
```
Comment (akshaymankar) [inline services/cannon/src/Cannon/Types.hs:122]: Why not generate it like in brig?

File: services/cargohold/src/CargoHold/Run.hs [hunk 10]
```diff
@@ -93,6 +97,15 @@ mkApp o = Codensity $ \k ->
...
         )
         r
+        cont
+
+    lookupReqId :: Logger -> Wai.Request -> IO RequestId
+    lookupReqId l r = case lookup requestIdName $ Wai.requestHeaders r of
+      Just rid -> pure $ RequestId rid
+      Nothing -> do
+        localRid <- RequestId . cs . UUID.toText <$> UUID.nextRandom
+        Log.info l $ "request-id" .= localRid ~~ "request" .= (show r) ~~ msg (val "generated a new request id for local request")
```
Comment (akshaymankar) [inline services/cargohold/src/CargoHold/Run.hs:107]: Same as brig, probably better to log individual items from the request rather than the whole request.

File: services/federator/src/Federator/ExternalServer.hs [hunk 11]
```diff
@@ -114,8 +116,8 @@ server ::
 server mgr intPort interpreter =
   API
     { status = Health.status mgr "internal server" intPort,
-      externalRequest = \component rpc remoteDomain remoteCert ->
-        Tagged $ \req respond -> runCodensity (interpreter (callInward component rpc remoteDomain remoteCert req)) respond
+      externalRequest = \component rpc mReqId remoteDomain remoteCert ->
+        Tagged $ \req respond -> runCodensity (interpreter (callInward component rpc (fromMaybe (RequestId "N/A") mReqId) remoteDomain remoteCert req)) respond
```
Comment (akshaymankar) [inline services/federator/src/Federator/ExternalServer.hs:120]: We should generate a request id here if we don't find one.

File: services/federator/src/Federator/InternalServer.hs [hunk 12]
```diff
@@ -89,8 +91,8 @@ server ::
 server mgr extPort interpreter =
   API
     { status = Health.status mgr "external server" extPort,
-      internalRequest = \remoteDomain component rpc ->
-        Tagged $ \req respond -> runCodensity (interpreter (callOutward remoteDomain component rpc req)) respond
+      internalRequest = \mReqId remoteDomain component rpc ->
+        Tagged $ \req respond -> runCodensity (interpreter (callOutward (fromMaybe (RequestId "N/A") mReqId) remoteDomain component rpc req)) respond
```
Comment (akshaymankar) [inline services/federator/src/Federator/InternalServer.hs:95]: Same as external server, we should generate a request id here.

File: services/galley/src/Galley/Run.hs [hunk 13]
```diff
@@ -106,24 +108,33 @@ mkApp opts =
...
+        cont
 
-    lookupReqId :: Request -> RequestId
-    lookupReqId = maybe def RequestId . lookup requestIdName . requestHeaders
+    lookupReqId :: Logger -> Request -> IO RequestId
+    lookupReqId l r = case lookup requestIdName $ requestHeaders r of
+      Just rid -> pure $ RequestId rid
+      Nothing -> do
+        localRid <- RequestId . cs . UUID.toByteString <$> UUID.nextRandom
+        Log.info l $ "request-id" .= localRid ~~ "request" .= (show r) ~~ msg (val "generated a new request id for local request")
```
Comment (akshaymankar) [inline services/galley/src/Galley/Run.hs:136]: Same as above.

File: services/gundeck/src/Gundeck/Monad.hs [hunk 14]
```diff
@@ -182,7 +181,7 @@ runDirect e m =
             )
 
 lookupReqId :: Request -> RequestId
-lookupReqId = maybe def RequestId . lookup requestIdName . requestHeaders
+lookupReqId = RequestId . fromMaybe "N/A" . lookup requestIdName . requestHeaders
```
Comment (akshaymankar) [inline services/gundeck/src/Gundeck/Monad.hs:184]: Same as cannon and federator, we should generate a req id here.

File: services/proxy/src/Proxy/Proxy.hs [hunk 15]
```diff
@@ -61,4 +60,4 @@ reqIdMsg = ("request" .=) . unRequestId
 {-# INLINE reqIdMsg #-}
 
 lookupReqId :: Request -> RequestId
-lookupReqId = maybe def RequestId . lookup requestIdName . requestHeaders
+lookupReqId = RequestId . fromMaybe "N/A" . lookup requestIdName . requestHeaders
```
Comment (akshaymankar) [inline services/proxy/src/Proxy/Proxy.hs:63]: We should generate a req id here.

File: tools/stern/src/Stern/App.hs [hunk 16]
```diff
@@ -71,7 +70,7 @@ newEnv :: Opts -> IO Env
 newEnv o = do
   mt <- Metrics.metrics
   l <- Log.mkLogger (O.logLevel o) (O.logNetStrings o) (O.logFormat o)
-  Env (mkRequest $ O.brig o) (mkRequest $ O.galley o) (mkRequest $ O.gundeck o) (mkRequest $ O.ibis o) (mkRequest $ O.galeb o) l mt def
+  Env (mkRequest $ O.brig o) (mkRequest $ O.galley o) (mkRequest $ O.gundeck o) (mkRequest $ O.ibis o) (mkRequest $ O.galeb o) l mt (RequestId "N/A")
```
Comment (akshaymankar) [inline tools/stern/src/Stern/App.hs:73]: I don't see this generated anywhere.

**Review by akshaymankar — CHANGES_REQUESTED** [review]
Comments inline. Two things I would consider blockers for merging:
1. The BackendNotification parser needs to be backwards compatible at least for a couple of releases.
2.  We shouldn't log the whole request, picking fields to log would be best.

File: libs/wai-utilities/src/Network/Wai/Utilities/Server.hs (hunk 4 above)
Comment (battermann) [reply libs/wai-utilities/src/Network/Wai/Utilities/Server.hs:364]: It's not, it sneaked in there somehow... good catch!

**Review by akshaymankar — APPROVED** [review]
